File: stuff.py
from hashlib import sha1
import sys

tree = "tree 6a0165c2aea6cfc5fba01029ede7a8da6c85f6f6"
author = "author Ben Thayer <ben@benthayer.com> 1599715620 -0500"
committer = "committer Ben Thayer <ben@benthayer.com> 1599715620 -0500"
message = "Initial commit"

def gen_hash(parent, salt):
    parent = 'parent ' + parent

    commit = '\n'.join([
        tree,
        parent,
        author,
        committer,
        '',
        message,
        '',
        salt,
        ''
    ])

    prefix = 'commit ' + str(len(commit)) + '\x00'
    obj = prefix + commit
    obj = obj.encode('utf-8')
    sha = sha1(obj).hexdigest()
    return sha

first_commit_hash = 'f9a17849fe28dff34647f698a392be2a9ce3617b'
print(first_commit_hash)
num_unique = int(sys.argv[1])
all_commits = set()
parent_hash = first_commit_hash
commit_hash = parent_hash
all_commits.add(parent_hash[:num_unique])

while len(all_commits) != 16**num_unique:
    salt = 0
    while commit_hash[:num_unique] in all_commits:
        salt += 1
        commit_hash = gen_hash(parent_hash, str(salt))
    print(commit_hash, salt)
    all_commits.add(commit_hash[:num_unique])
    parent_hash = commit_hash

File: stuff.go
package main

import "crypto/sha1"
import "encoding/hex"
import "encoding/binary"

import "fmt"
import "strings"
import "strconv"

import "os"
import "bufio"

var tree = "tree 6a0165c2aea6cfc5fba01029ede7a8da6c85f6f6"
var author = "author Ben Thayer <ben@benthayer.com> 1599715620 -0500"
var committer = "committer Ben Thayer <ben@benthayer.com> 1599715620 -0500"
var message = "Initial commit"

func gen_hash(parent_hash string, salt string) [20]byte {
    parent := "parent " + parent_hash
    commit := strings.Join(
        []string{
            tree,
            parent,
            author,
            committer,
            "",
            message,
            "",
            salt,
            ""},
        "\n")

    prefix := "commit " + strconv.Itoa(len(commit)) + "\x00"
    object := prefix + commit
    data := []byte(object)
    shasum := sha1.Sum(data)
    return shasum
}


func check(e error) {
    if e != nil {
        panic(e)
    }
}


func main() {
    f, err := os.Create("./commits")
    check(err)

    defer f.Close()

    w := bufio.NewWriter(f)

    // 1<<28 == 16^7
    const num_total_hashes = 1<<28
    var commit_generated [num_total_hashes]bool

    parent_hash := "f9a17849fe28dff34647f698a392be2a9ce3617b"
    hash_decoded, err := hex.DecodeString(parent_hash[:8])
    check(err)
    int_hash := binary.BigEndian.Uint32(hash_decoded) / 16
    commit_generated[int_hash] = true

    var expected float64 = 0
    var progress float64 = 0

    for commit_number := 0; commit_number < (num_total_hashes); commit_number++ {
        expected += float64(num_total_hashes)/float64(num_total_hashes-commit_number)
    }

    var shasum [20]byte
    for commit_number := 1; commit_number < (num_total_hashes); commit_number++ {
        salt := 0
        for commit_generated[int_hash] {
            salt++
            shasum = gen_hash(parent_hash, strconv.Itoa(salt))
            hash := make([]byte, 4)
            copy(hash, shasum[:4])
            hash[3] = hash[3] - (hash[3] % 16)
            int_hash = binary.BigEndian.Uint32(hash) / 16
        }
        commit_generated[int_hash] = true
        parent_hash = hex.EncodeToString(shasum[:])
        w.WriteString(parent_hash + " " + strconv.Itoa(salt) + "\n")
        progress += float64(num_total_hashes)/float64(num_total_hashes-commit_number)
        if commit_number > (num_total_hashes - 12000) {
            w.Flush()
        }
        fmt.Print(strconv.FormatFloat(progress/expected, 'f', 8, 64) + "   " + strconv.Itoa(commit_number) + "\r")
    }
}
